types/wlr_output.c

```diff
@@ -188,94 +188,172 @@ void wlr_output_set_position(struct wlr_output *output, int32_t lx,
 	}
 }
 
-static bool set_cursor(struct wlr_output *output, const uint8_t *buf,
-		int32_t stride, uint32_t width, uint32_t height, int32_t hotspot_x,
-		int32_t hotspot_y) {
-	if (output->impl->set_cursor
-			&& output->impl->set_cursor(output, buf, stride, width, height,
-				hotspot_x, hotspot_y, true)) {
-		output->cursor.is_sw = false;
-		return true;
+void wlr_output_init(struct wlr_output *output, struct wlr_backend *backend,
+		const struct wlr_output_impl *impl) {
+	output->backend = backend;
+	output->impl = impl;
+	wl_list_init(&output->modes);
+	output->transform = WL_OUTPUT_TRANSFORM_NORMAL;
+	output->scale = 1;
+	wl_list_init(&output->cursors);
+	wl_signal_init(&output->events.frame);
+	wl_signal_init(&output->events.swap_buffers);
+	wl_signal_init(&output->events.resolution);
+	wl_signal_init(&output->events.destroy);
+}
+
+void wlr_output_destroy(struct wlr_output *output) {
+	if (!output) {
+		return;
 	}
 
-	wlr_log(L_INFO, "Falling back to software cursor");
+	wl_signal_emit(&output->events.destroy, output);
 
-	output->cursor.is_sw = true;
-	output->cursor.width = width;
-	output->cursor.height = height;
+	struct wlr_output_mode *mode, *tmp_mode;
+	wl_list_for_each_safe(mode, tmp_mode, &output->modes, link) {
+		free(mode);
+	}
+	wl_list_remove(&output->modes);
+	if (output->impl && output->impl->destroy) {
+		output->impl->destroy(output);
+	} else {
+		free(output);
+	}
+}
 
-	if (!output->cursor.renderer) {
-		output->cursor.renderer = wlr_gles2_renderer_create(output->backend);
-		if (!output->cursor.renderer) {
-			return false;
-		}
+void wlr_output_effective_resolution(struct wlr_output *output,
+		int *width, int *height) {
+	// TODO: Scale factor
+	if (output->transform % 2 == 1) {
+		*width = output->height;
+		*height = output->width;
+	} else {
+		*width = output->width;
+		*height = output->height;
 	}
+}
 
-	if (!output->cursor.texture) {
-		output->cursor.texture =
-			wlr_render_texture_create(output->cursor.renderer);
-		if (!output->cursor.texture) {
-			return false;
-		}
+void wlr_output_make_current(struct wlr_output *output) {
+	output->impl->make_current(output);
+}
+
+static void output_cursor_render(struct wlr_output_cursor *cursor) {
+	struct wlr_texture *texture = cursor->texture;
+	struct wlr_renderer *renderer = cursor->renderer;
+	if (cursor->surface != NULL) {
+		texture = cursor->surface->texture;
+		renderer = cursor->surface->renderer;
 	}
 
-	return wlr_texture_upload_pixels(output->cursor.texture,
-		WL_SHM_FORMAT_ARGB8888, stride, width, height, buf);
+	// We check texture->valid because some clients set a cursor surface
+	// with a NULL buffer to hide it
+	if (renderer && texture && texture->valid) {
```

Comment:
I think we are going to start using `wlr_surface_has_buffer()` for this which is future proof if texture->valid goes away.

Reply:
Good point.